execute/executor.go

```diff
@@ -173,75 +172,127 @@ func (v *createExecutionNodeVisitor) Visit(node plan.Node) error {
 		}
 	}
 
-	// Build execution context
-	ec := executionContext{
-		es:            v.es,
-		parents:       make([]DatasetID, len(node.Predecessors())),
-		streamContext: streamContext,
+	// There are three types of instantiations we need to support here, and for
+	// each one of these, there are source nodes and non-source nodes.
+	//
+	// 1. Standard instantiation. These are non-parallel, non-merge nodes.
+	//    There is a 1:1 relationship between planner node and execution graph
+	//    node. Only a single copy of the node is made and it references the
+	//    single copy of the predecessor nodes.
+	//
+	// 2. Parallel instantiation. There are multiple copies of the node and of
+	//    all predecessors. This applies to both source and non-source nodes.
+	//
+	// 3. Merge instantiation. There is a single copy of the node, but multiple copies of the
+	//    predecessors. These copies merge into the node.
+
+	copies := 1
+	if attr, ok := ppn.OutputAttrs[plan.ParallelRunKey]; ok {
+		copies = attr.(plan.ParallelRunAttribute).Factor
 	}
 
-	for i, pred := range nonYieldPredecessors(node) {
-		ec.parents[i] = DatasetIDFromNodeID(pred.ID())
+	predCopies := 1
+	if attr, ok := ppn.OutputAttrs[plan.ParallelMergeKey]; ok {
+		predCopies = attr.(plan.ParallelMergeAttribute).Factor
 	}
 
+	// Build execution context for each copy.
+	ec := make([]executionContext, copies)
+	for i := 0; i < copies; i++ {
+		ec[i] = executionContext{
+			es:            v.es,
+			parents:       make([]DatasetID, len(node.Predecessors())*predCopies),
+			streamContext: streamContext,
+			parallelOpts:  ParallelOpts{Group: i, Factor: copies},
+		}
+
+		for pi, pred := range nonYieldPredecessors(node) {
+			for j := 0; j < predCopies; j++ {
+				ec[i].parents[pi*predCopies+j] = DatasetIDFromNodeID(pred.ID(), j)
```

Comment:
This code is not so easy to follow. I really thought hard about the pros and cons of cloning and modifying it for handling the different cases. For better or worse, chose this route ...

The thing to remember here is that if `predCopies` is > `1` and that inner loop at 210 is more than just a single iteration, it means we are handling the merge node. The `i` var will always be zero in that case because the merge node is not copied. It must handle the incoming datasets from all copies of the predecessors, so that's why we supply `j`.

Reply:
This code makes sense to me for the case of the merge node, but for a parallel node that's not a source, I'm not seeing how it works right. Won't this result in every instantiation of a parallel node getting the data set from the 0th instantiation of its predecessor?

Reply:
You're right. There is a problem. I think it needs to be `i+j`. It's not getting easier to follow.

This would mean that in my testing the incoming dataset id was not correct for the `to()` nodes. I wonder why it all appeared to work correctly? Maybe `to()` doesn't depend on a correct dataset id?

Reply:
So it seems the call to `newConsecutiveTransport` is the official linking operation. The `parents` struct in the execution context is just an informational type of thing for transformations to use. At least that's how I'm currently understanding things.

Reply:
Looks like the only kinds of transformations making use of `parents` are `union`, `join`, `diff`, etc.  So this explains why it didn't show up in testing. I'm going to look for some way to verify that parents is indeed correct.

Reply:
Adding some log messages confirms we need to use `i+j` in order to ensure that `parents` is correct for both parallel-run and parallel-merge nodes.